tests/load/locustfiles/locustfile.py

```diff
@@ -31,119 +48,392 @@ def _(parser: Any):
+            data=data,
+            headers=self.REST_HEADERS,
+            catch_response=True,
+        ) as response:
+            if response.status_code == 0:
+                raise ZeroStatusRequestError()
+            if response.status_code != 201:
+                response.failure(f"{response.status_code=}, expected 201, {response.text=}")
+                return
+
+    def recv_hello(self, data: str, recv_time: float) -> None:
+        """Verify contents of 'hello' message and report response statistics to Locust.
+
+        Args:
+            data: utf-8 data received from the server
+            recv_time: Time when data was received
+        """
+        response_time: float = 0
+        exception: str | None = None
+
+        try:
+            message = HelloMessage(**json.loads(data))
+
+            # If a user is not subscribed to any channels upon re-connection a new
+            # ID will be assigned
+            self.uaid = message.uaid
+
+            # Get last 'hello' message record
+            record: MessageRecord = next(
+                (
+                    self.message_records[i]
+                    for i in reversed(range(len(self.message_records)))
+                    if self.message_records[i].message_type == message.messageType
+                ),
+                None,
+            )
+            if record:
+                response_time = (recv_time - record.send_time) * 1000
+            else:
+                exception = f"There is no record of the '{message.messageType}' message"
+                logger.error(f"{exception}. Contents: {message}")
+        except (ValidationError, JSONDecodeError) as error:
+            exception = str(error)
+
+        self.environment.events.request.fire(
+            request_type="WSS",
+            name="hello - recv",
+            response_time=response_time,
+            response_length=len(data.encode("utf-8")),
+            exception=exception,
+            context=self.context(),
+        )
+
+    def recv_notification(self, ws: WebSocketApp, data: str, recv_time: float) -> None:
+        """Verify contents of 'notification' message, report response statistics to Locust and
+        send 'ack'.
+
+        Args:
+            ws: WebSocket class object
+            data: utf-8 data received from the server
+            recv_time: Time when data was received
+        """
+        message: NotificationMessage | None = None
+        response_time: float = 0
+        exception: str | None = None
+
+        try:
+            message = NotificationMessage(**json.loads(data))
+            decoded_data: str = base64.urlsafe_b64decode(message.data + "===").decode("utf8")
+
+            record: MessageRecord = next(
+                (
+                    r
+                    for r in self.message_records
+                    if r.message_type == message.messageType and r.data == decoded_data
+                ),
+                None,
+            )
```

Comment:
Does it make sense to "pop" a recv'd message out of `message_records`? Otherwise, if the server just returned the very first message it got, wouldn't this still return back that things were successful?

(Alternately, since this is a load test and not an integration or unit test, if you wanted to save a bit of memory, maybe have the data always be the same and validate that the received notification data matches expected?)

Reply:
Originally I was poping messages from message_records. However WebSocket connection issues before or during 'ack' would happen (async yay!) and then when the message was recent, we had no matching message in message_records. I've opted to store all the messages we sent and am identifying them with the 'decoded_data", which i'm counting on being unique. 

Verifying that message data is correct is nice, but what I'm really after is measuring the response time.

You are right about there being a memory impact though.